Description: Add FrameAttached notifications to WebContentsObserver, so that we can track frames
Author: Chris Coulson <chris.coulson@canonical.com>

diff --git a/content/browser/web_contents/web_contents_impl.cc b/content/browser/web_contents/web_contents_impl.cc
--- a/content/browser/web_contents/web_contents_impl.cc
+++ b/content/browser/web_contents/web_contents_impl.cc
@@ -2587,16 +2587,19 @@ FrameTreeNode* WebContentsImpl::FindFram
 
   return NULL;
 }
 
 void WebContentsImpl::OnFrameAttached(
     int64 parent_frame_id,
     int64 frame_id,
     const std::string& frame_name) {
+  FOR_EACH_OBSERVER(WebContentsObserver, observers_,
+                    FrameAttached(message_source_, parent_frame_id, frame_id));
+
   FrameTreeNode* parent = FindFrameTreeNodeByID(parent_frame_id);
   if (!parent)
     return;
 
   FrameTreeNode* node = new FrameTreeNode(frame_id, frame_name);
   parent->AddChild(node);
 }
 
diff --git a/content/public/browser/web_contents_observer.h b/content/public/browser/web_contents_observer.h
--- a/content/public/browser/web_contents_observer.h
+++ b/content/public/browser/web_contents_observer.h
@@ -195,16 +195,19 @@ class CONTENT_EXPORT WebContentsObserver
   // place.
   virtual void DidOpenRequestedURL(WebContents* new_contents,
                                    const GURL& url,
                                    const Referrer& referrer,
                                    WindowOpenDisposition disposition,
                                    PageTransition transition,
                                    int64 source_frame_id) {}
 
+  virtual void FrameAttached(RenderViewHost* render_view_host,
+                             int64 parent_frame_id,
+                             int64 frame_id) {}
   virtual void FrameDetached(RenderViewHost* render_view_host,
                              int64 frame_id) {}
 
   // These two methods correspond to the points in time when the spinner of the
   // tab starts and stops spinning.
   virtual void DidStartLoading(RenderViewHost* render_view_host) {}
   virtual void DidStopLoading(RenderViewHost* render_view_host) {}
 
